frontend/src/app/containers/ExperimentPage.js

```diff
@@ -648,20 +649,57 @@ export class ExperimentDetail extends React.Component {
     // Ignore subsequent clicks if already in progress
     if (isEnabling) { return; }
 
+    let installAddonPromise = null;
+    if (!this.props.hasAddon) {
+      const { installAddon, requireRestart, eventCategory } = this.props;
+
+      installAddonPromise = installAddon(
+        requireRestart,
+        sendToGA,
+        eventCategory);
+    }
+    let progressButtonWidth;
+    if (this.props.hasAddon) {
+      progressButtonWidth = evt.target.offsetWidth;
+    }
+
     this.setState({
       isEnabling: true,
       isDisabling: false,
       shouldShowTourDialog: true,
-      progressButtonWidth: evt.target.offsetWidth
+      progressButtonWidth
     });
 
-    enableExperiment(experiment);
+    function finishEnabling() {
+      enableExperiment(experiment);
 
-    sendToGA('event', {
-      eventCategory: 'ExperimentDetailsPage Interactions',
-      eventAction: 'Enable Experiment',
-      eventLabel: experiment.title
-    });
+      sendToGA('event', {
+        eventCategory: 'ExperimentDetailsPage Interactions',
+        eventAction: 'Enable Experiment',
+        eventLabel: experiment.title
+      });
+    }
+
+    if (installAddonPromise === null) {
+      finishEnabling();
+      return;
+    }
+
+    installAddonPromise.then(() => {
+      return new Promise((resolve, reject) => {
+        let i = 0;
+        const interval = setInterval(() => {
+          i++;
+          if (window.navigator.testpilotAddon) {
```

Comment:
This looks good. In the future it'd be interesting to see if we can get away with removing this flag and use `mozAddonManager.getAddonByID('@testpilot-addon')`